File: demos/java_demos/src/main/java/com/orientwalt/demos/HtdfTransfer.java
/*
 * author: yqq
 * date: 2020-10-09
 * descriptions: htdf transfer demo  ( java )
 *
 * reference: ../demo/python/htdf_transfer.py
 *
 */

package com.orientwalt.demos;


import com.alibaba.fastjson.JSONObject;
import com.google.common.primitives.Bytes;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import org.bouncycastle.jce.ECNamedCurveTable;
import org.bouncycastle.jce.spec.ECNamedCurveParameterSpec;
import org.bouncycastle.math.ec.ECPoint;
import org.bouncycastle.util.encoders.Hex;
import org.web3j.crypto.ECKeyPair;
import org.web3j.crypto.Sign;

import java.io.IOException;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;



public class HtdfTransfer {

    private static final String gChainId = "testchain";        // mainchain or testchain
    private static final String gFromAddress = "htdf1xwpsq6yqx0zy6grygy7s395e2646wggufqndml";
    private static final String gPrivateKey = "279bdcd8dccec91f9e079894da33d6888c0f9ef466c0b200921a1bf1ea7d86e8";
    private static final String gNodeIpPort = "htdf2020-test01.orientwalt.cn:1317";
    private static final int gGasLimit = 30000;
    private static final int gGasPrice =  100;


    // mainchain parameters
//    private static final String gChainId = "mainchain";                                     // mainchain or testchain
//    private static final String  gFromAddress = "YOUR_HTDF_ADDRESS"                         //TODO
//    private static final String  gPrivateKey = "YOUR_HTDF_ADDRESS_PRIVATE_KEY"              //TODO
//    private static final String  gNodeIpPort = "YOUR_HTDF_FULL_NODE_IP_ADDRESS:1317" 		  //TODO
//    private static final int gGasLimit = 30000;
//    private static final int gGasPrice =  100;


    private static final String  gUnsignedStr = "{\n" +
            "    \"account_number\": \"%d\",\n" +
            "    \"chain_id\": \"%s\",\n" +
            "    \"fee\": {\n" +
            "        \"gas_price\": \"%d\",\n" +
            "        \"gas_wanted\": \"%d\"\n" +
            "    },\n" +
            "    \"memo\": \"%s\",\n" +
            "    \"msgs\": [{\n" +
            "        \"Amount\": [{\n" +
            "        \"amount\": \"%d\",\n" +
            "        \"denom\": \"satoshi\"\n" +
            "    }],\n" +
            "        \"Data\": \"\",\n" +
            "        \"From\": \"%s\",\n" +
            "        \"GasPrice\": %d,\n" +
            "        \"GasWanted\": %d,\n" +
            "        \"To\": \"%s\"\n" +
            "    }],\n" +
            "    \"sequence\": \"%d\"\n" +
            "}";

    private  static final  String gBroadcastStr = "{\n" +
            "    \"type\": \"auth/StdTx\",\n" +
            "        \"value\":{\n" +
            "        \"msg\": [{\n" +
            "            \"type\": \"htdfservice/send\",\n" +
            "            \"value\":{\n" +
            "                \"From\": \"%s\",\n" +
            "                \"To\": \"%s\",\n" +
            "                \"Amount\": [{\n" +
            "                    \"denom\": \"satoshi\",\n" +
            "                    \"amount\": \"%d\"\n" +
            "                }],\n" +
            "                \"Data\": \"\",\n" +
            "                \"GasPrice\": \"%d\",\n" +
            "                \"GasWanted\": \"%d\"\n" +
            "            }\n" +
            "        }],\n" +
            "    \"fee\": {\n" +
            "        \"gas_wanted\": \"%d\",\n" +
            "        \"gas_price\": \"%d\"\n" +
            "    },\n" +
            "    \"signatures\": [{\n" +
            "        \"pub_key\": {\n" +
            "        \"type\": \"tendermint/PubKeySecp256k1\",\n" +
            "        \"value\": \"%s\"\n" +
            "    },\n" +
            "    \"signature\": \"%s\"\n" +
            "    }],\n" +
            "    \"memo\": \"%s\"\n" +
            "    }\n" +
            "}";

    // this method exists a bug. when s[0] is greater than 0x80 , R is right but S is wrong.
    // I have not found solution yet, so I use web3j's sign to replace this method. 2020-10-09
    // public static byte[] sign(byte[] message, byte[] privateKey) {
    //     Security.addProvider(new BouncyCastleProvider());

    //     final String SECP256K1 = "secp256k1";
    //     ECNamedCurveParameterSpec spec = ECNamedCurveTable.getParameterSpec(SECP256K1);

    //     ECDSASigner ecdsaSigner = new ECDSASigner(new HMacDSAKCalculator(new SHA256Digest()));

    //     ECDomainParameters domain = new ECDomainParameters(spec.getCurve(),  spec.getG(),  spec.getN());

    //     ECPrivateKeyParameters privateKeyParams =
    //             new ECPrivateKeyParameters(new BigInteger(1, privateKey), domain);

    //     ecdsaSigner.init(true, privateKeyParams);

    //     BigInteger[] sig = ecdsaSigner.generateSignature(message);

    //     String  R = sig[0].toString(16);
    //     String  S = sig[1].toString(16);
    //     return Hex.decode(R + S);
    // }

    // use web3j
    public static byte[] sign(byte[] message, byte[] privateKey) {
        ECKeyPair eckKeyPair = ECKeyPair.create(privateKey);
        Sign.SignatureData sig = Sign.signMessage(message, eckKeyPair, false);
        return Bytes.concat(sig.getR(), sig.getS());
    }

    public static BigInteger HtdfToSatoshi(BigDecimal amount) {
        // 1 HTDF = 10 ^ 8 satoshi
        return amount.multiply(new BigDecimal(10).pow(8)).toBigInteger();
    }

    public static String doHttpGet(String url) throws Exception {
        CloseableHttpClient httpClient = HttpClientBuilder.create().build();
        HttpGet httpGet = new HttpGet(url);

        String ret = "";
        try {
            CloseableHttpResponse response = httpClient.execute(httpGet);
            if (response.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
                HttpEntity responseEntity = response.getEntity();

                if (responseEntity != null) {
                    ret = EntityUtils.toString(responseEntity);
                    response.close();
                }
            } else {
                throw new Exception(response.getStatusLine().getReasonPhrase());
            }

        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            httpClient.close();
        }
        return ret;
    }

    public static JSONObject doHttpPost(String url, JSONObject jsonObject) throws RuntimeException {
        HttpClient client = HttpClientBuilder.create().build();
        HttpPost post = new HttpPost(url);
        JSONObject response = null;

        try {
            StringEntity s = new StringEntity(jsonObject.toString());
            s.setContentEncoding("UTF-8");
            s.setContentType("application/json");
            post.setEntity(s);
            HttpResponse res = client.execute(post);
            if (res.getStatusLine().getStatusCode() == HttpStatus.SC_OK) {
                HttpEntity entity = res.getEntity();
                String result = EntityUtils.toString(entity);
                response = JSONObject.parseObject(result);
            } else {
                throw new Exception(res.getStatusLine().getReasonPhrase());
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        return response;
    }

    public static byte[] getSha256(byte[] data) throws Exception {
        MessageDigest messageDigest = MessageDigest.getInstance("SHA-256");
        messageDigest.update(data);
        return messageDigest.digest();
    }

    public static byte[] generatePublicKey(byte[] privateKey) {
        ECNamedCurveParameterSpec spec = ECNamedCurveTable.getParameterSpec("secp256k1");
        ECPoint pointQ = spec.getG().multiply(new BigInteger(1, privateKey));
        return pointQ.getEncoded(true);
    }

    public static void transfer(String dstAddress, BigInteger amount, String memo) throws Exception {

        if (amount.compareTo(BigInteger.ONE) <= 0
                || amount.compareTo(HtdfToSatoshi(BigDecimal.valueOf(6000_0000))) >= 0) {
            throw new Exception("invalid amount");
        }

        String reqUrl = String.format("http://%s/auth/accounts/%s", gNodeIpPort, gFromAddress);
        String rsp = doHttpGet(reqUrl);
        JSONObject accInfo = JSONObject.parseObject(rsp);

        JSONObject value = accInfo.getJSONObject("value");

        double balance = Double.parseDouble(value.getJSONArray("coins").getJSONObject(0).getString("amount"));
        if (HtdfToSatoshi(BigDecimal.valueOf(balance)).compareTo(amount) < 0) {
            throw new Exception("balance is not enough for transfer");
        }

        int accountNumber = Integer.parseInt(value.getString("account_number"));
        int sequence = Integer.parseInt(value.getString("sequence"));

        String unsignedStr = gUnsignedStr.replaceAll("[\\t\\n\\r ]", "");
        String fmtUnsignedStr = String.format(unsignedStr, accountNumber, gChainId, gGasPrice, gGasLimit, memo, amount,
                gFromAddress, gGasPrice, gGasLimit, dstAddress, sequence);
        System.out.println("formatted raw transaction string: " + fmtUnsignedStr);

        byte[] sha256Bytes = getSha256(fmtUnsignedStr.getBytes(StandardCharsets.UTF_8));
        System.out.println("msgHash : " + Hex.toHexString(sha256Bytes));

        byte[] privKeyBytes = Hex.decode(gPrivateKey);
        byte[] sigRS = sign(sha256Bytes, privKeyBytes);
        System.out.println("signature : " + Hex.toHexString(sigRS));

        final Base64.Encoder base64 = Base64.getEncoder();
        String b64Sig = base64.encodeToString(sigRS);
        System.out.println("base64 encoded signature : " + b64Sig);

        byte[] pubKeyBytes = generatePublicKey(privKeyBytes);
        String b64PubKey = base64.encodeToString(pubKeyBytes);
        System.out.println("base64 encoded public key: " + b64PubKey);

        String broadcastStr = gBroadcastStr.replaceAll("[\\t\\n\\r ]", "");
        String fmtBroadcastStr = String.format(broadcastStr, gFromAddress, dstAddress, amount, gGasPrice, gGasLimit,
                gGasLimit, gGasPrice, b64PubKey, b64Sig, memo);

        System.out.println("formated broadcast string : " + fmtBroadcastStr);

        String hexBroadcastStr = Hex.toHexString(fmtBroadcastStr.getBytes());
        System.out.println("hex broadcast str: " + hexBroadcastStr);

        JSONObject reqObj = new JSONObject();
        reqObj.put("tx", hexBroadcastStr);

        String broadcastUrl = String.format("http://%s/hs/broadcast", gNodeIpPort);
        JSONObject bcastResponse = doHttpPost(broadcastUrl, reqObj);
        System.out.println("broadcast response : " + bcastResponse.toString());

        if (bcastResponse.containsKey("raw_log")) {
            throw new Exception(bcastResponse.getString("raw_log"));
        }

        System.out.println("================");
        String txhash = bcastResponse.getString("txhash");
        String tx_url = "http://htdf2020-test01.orientwalt.cn:1317/txs/" + txhash;
        System.out.println("tx_url: " + tx_url);

        int nSecondsToWait = 10;
        System.out.printf("Waiting %d seconds for tx ", nSecondsToWait);
        for (int i = nSecondsToWait; i > 0; i--) {
            System.out.printf(".");
            Thread.sleep(1000);
        }
        System.out.println();

        String strTx = doHttpGet(tx_url);
        // JSONObject.parseObject(strTx);
        System.out.println(strTx);

        // Congratulation! If everything is ok.
        // then, you could get transaction details on blockchain by http RPC (testchain)
        // curl http://htdf2020-test01.orientwalt.cn:1317/txs/{txhash}
        // example: curl
        // http://htdf2020-test01.orientwalt.cn:1317/txs/B85D5829299C63BC38BB2F215ED1405EEAB79982E14C3FEC4A748EC9A1604A57

    }

    public static void main(String[] args) throws Exception {

        String dstAddress = "htdf1jrh6kxrcr0fd8gfgdwna8yyr9tkt99ggmz9ja2";
        String memo = "htdf transaction demo in java";
        BigInteger amountSatoshi = HtdfToSatoshi(BigDecimal.valueOf(0.12345678));
        transfer(dstAddress, amountSatoshi, memo);

    }

}

File: demos/python_demos/htdf_transfer.py
#!coding:utf8
# author:yqq
# date:2019/5/6 0006 18:07   create demo
# date:2019/9/29 0029 11:03  python2 -> python3
# date: 2020/10/10   refactor


import json
import hashlib
import time
import ecdsa
import traceback
import coincurve
import base64
import requests
import logging
from typing import Tuple, Dict
from binascii import hexlify, unhexlify
from bech32 import bech32_decode
from func_timeout import func_set_timeout

#  configurations for testchain
g_chainid = 'testchain'
g_fromaddr = 'htdf1xwpsq6yqx0zy6grygy7s395e2646wggufqndml'
g_privkey = '279bdcd8dccec91f9e079894da33d6888c0f9ef466c0b200921a1bf1ea7d86e8'
g_node_ip_port = 'htdf2020-test01.orientwalt.cn:1317'
# g_node_ip_port = '192.168.0.171:1317'
g_gaslimit = 30000
g_gasprice = 100

# TODO: configurations for mainchain
# g_chainid = 'mainchain'
# g_strFrom = 'YOUR_HTDF_ADDRESS'
# g_PrivKey = 'YOUR_HTDF_ADDRESS_PRIVATE_KEY'
# g_node_ip_port = 'YOUR_HTDF_FULL_NODE_IP_ADDRESS:1317'
# g_gaslimit = 30000
# g_gasprice = 100


g_unsigned_rawtx = """{\
    "account_number": "%d",\
	"chain_id": "%s",\
	"fee": {\
			"gas_price": "%d",\
			"gas_wanted": "%d"\
	},\
    "memo": "%s",\
	"msgs": [{\
		"Amount": [{\
			"amount": "%d",\
            "denom": "satoshi"\
		}],\
        "Data": "",\
        "From": "%s",\
        "GasPrice": %s,\
        "GasWanted": %s,\
		"To": "%s"\
	}],\
    "sequence": "%d"\
    }"""

g_broadcast_str = """{
        "type": "auth/StdTx",
        "value":{
            "msg": [{
                "type": "htdfservice/send",  
                "value":{
                    "From": "%s",
                    "To": "%s",
                    "Amount": [{
                        "denom": "satoshi",
                        "amount": "%d"
                    }],
                    "Data": "",
                    "GasPrice": "%d",
                    "GasWanted": "%d"
                }
            }],
            "fee": {
                "gas_wanted": "%d",
                "gas_price": "%d"
            },
            "signatures": [{
                "pub_key": {
                    "type": "tendermint/PubKeySecp256k1",
                    "value": "%s"
                },
                "signature": "%s"
            }],
            "memo": "%s"
        }
    }"""


def sign(rawhash, key):
    if coincurve and hasattr(coincurve, 'PrivateKey'):
        pk = coincurve.PrivateKey(key)
        signature = pk.sign_recoverable(rawhash, hasher=None)
        # v = safe_ord(signature[64]) + 27
        r = signature[0:32]
        s = signature[32:64]
        return r, s

def privkey_to_pubkey(privkey: str) -> str:
    sk = ecdsa.SigningKey.from_string(unhexlify(privkey), curve=ecdsa.SECP256k1)
    s = sk.get_verifying_key().to_string(encoding='compressed')
    return hexlify(s).decode('latin')


def get_account_info(address: str) -> Tuple[int, int]:
    url = 'http://{0}/auth/accounts/{1}'.format(g_node_ip_port.strip(), address.strip())
    rsp = requests.get(url)

    if rsp.status_code != 200:
        if rsp.status_code == 204:
            raise Exception('not found any account info ')
        raise Exception('get account info error: {}'.format(rsp.status_code))

    rsp = rsp.json()
    account_number = int(rsp['value']['account_number'])
    sequence = int(rsp['value']['sequence'])
    return account_number, sequence


def is_valid_address(address: str) -> bool:
    if not (len(address) == 43 and address.islower()):
        return False

    prefix, data = bech32_decode(address)
    if prefix is None or data is None:
        return False

    if prefix == 'htdf' and len(data) == 32:
        return True

    return False


def htdf_to_satoshi(amount_htdf: float) -> int:
    return int(amount_htdf * 10 ** 8)


def broadcast_tx(json_data: dict) -> str:
    assert 'tx' in json_data, 'invalid json_data to broadcast'

    post_data = json.dumps(json_data)
    url = 'http://%s/hs/broadcast' % (g_node_ip_port)
    rsp = requests.post(url, post_data)

    if rsp.status_code != 200:
        raise Exception(rsp.text)

    rsp_json = rsp.json()
    if 'code' in rsp_json or 'raw_log' in rsp_json:
        raise Exception(rsp.text)

    txhash = str(rsp_json['txhash'])

    return txhash


def transfer(from_address: str, to_address: str, amount_satoshi: int, memo: str) -> str:
    # step 1 : arguments check
    if not is_valid_address(from_address):
        raise Exception("invalid from_address address : {}".format(from_address))

    if not is_valid_address(to_address):
        raise Exception("invalid to_address address : {}".format(to_address))

    if amount_satoshi < 1 or amount_satoshi > htdf_to_satoshi(6000_0000):
        raise Exception("amount_satoshi {} is invalid".format(amount_satoshi))

    # step 2 : get account details
    account_number, sequence = get_account_info(from_address)
    logging.info('account_number : {}, sequence: {} '.format(account_number, sequence))

    # step 3 : format raw transaction
    fmt_unsigned_txstr = g_unsigned_rawtx.replace(' ', '').replace('\t', '').replace('\n', '')
    fmt_unsigned_txstr = fmt_unsigned_txstr % (
        account_number, g_chainid, g_gasprice, g_gaslimit, memo,
        amount_satoshi, from_address, g_gasprice, g_gaslimit,
        to_address, sequence)

    logging.info("formatted raw transaction str: {}".format(fmt_unsigned_txstr))

    # step 4 : make signature
    shadata = hashlib.sha256(fmt_unsigned_txstr.encode('utf-8')).digest()
    logging.info("sha256(fmt_unsigned_txstr): {}".format(hexlify(shadata)))

    r, s = sign(shadata, unhexlify(g_privkey))
    logging.info('r:' + hexlify(r).decode(encoding='utf8'))
    logging.info('s:' + hexlify(s).decode(encoding='utf8'))

    b64sig = base64.b64encode(r + s).decode(encoding='utf8')
    logging.info("base64encode(signature) : {}".format(b64sig))

    # step 5 : format broadcast string
    pubkey = privkey_to_pubkey(privkey=g_privkey)
    b64pubkey = base64.b64encode(unhexlify(pubkey)).decode(encoding='utf8')
    logging.info("base64encode(public key) :" + b64pubkey)

    fmt_broadcast_str = g_broadcast_str.replace(' ', '').replace('\n', '').replace('\t', '')
    fmt_broadcast_str = fmt_broadcast_str % (
        from_address, to_address, amount_satoshi,
        g_gasprice, g_gaslimit, g_gaslimit, g_gasprice, b64pubkey, b64sig, memo)
    logging.info("broadcast str: {}".format(fmt_broadcast_str))

    broadcast_data = hexlify(bytes(fmt_broadcast_str, encoding='utf8')).decode(encoding='utf8')
    broadcast_data = {'tx': broadcast_data}
    print(broadcast_data)

    # step 6 : broadcast
    # tx_hash = broadcast_tx(broadcast_data)
    # logging.info("txhash : {}".format(tx_hash))
    # return tx_hash


@func_set_timeout(30)
def get_tranaction(tx_hash: str) -> Dict:
    while True:
        url = 'http://{0}/txs/{1}'.format(g_node_ip_port.strip(), tx_hash.strip())
        rsp = requests.get(url)
        if rsp.status_code == 404:
            time.sleep(3)
            continue
        tx = rsp.json()
        return tx
    pass


if __name__ == '__main__':

    logging.basicConfig(level=logging.INFO,
                        format='[%(asctime)s |  %(levelname)s | %(filename)s |%(funcName)s:%(lineno)d] %(message)s')

    try:
        from_address = g_fromaddr
        to_address = 'htdf1jrh6kxrcr0fd8gfgdwna8yyr9tkt99ggmz9ja2'
        amount_satoshi = htdf_to_satoshi(0.12345678)
        memo = 'transaction demo in python'

        tx_hash = transfer(from_address, to_address, amount_satoshi, memo)

        # get transaction details
        tx = get_tranaction(tx_hash=tx_hash)
        logging.info(json.dumps(tx, indent=4))

    except Exception as e:
        traceback.print_exc()

    pass
